### PR TITLE
Validate text selection before updating it inside of EditableText

### LINKED ISSUE #35848: invalid text selection: `TextSelection` with `TextEditingController`
Probably related to this: https://github.com/flutter/flutter/issues/17647

## Steps to Reproduce
1. Run app
2. Tap on first TextField and type anything (a simple 'a' works).
3. Tap on second TextField (first one will loose focus)
4. Tap on X (clear) button from first TextField, something really weird will happen and an error will be logged.
A buggy ripple will happen and cover the full TextField:
image

Source code:
```dart
import 'package:flutter/material.dart';

final TextEditingController _textController1 = TextEditingController();
final TextEditingController _textController2 = TextEditingController();

class MainPage extends StatelessWidget {
  MainPage({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Column(
          children: <Widget>[
            SearchField(textController: _textController1),
            SearchField(textController: _textController2),
          ],
        ),
      ),
    );
  }
}

class SearchField extends StatefulWidget {
  final TextEditingController textController;

  SearchField({this.textController});

  @override
  _SearchFieldState createState() => _SearchFieldState();
}

class _SearchFieldState extends State<SearchField> {
  @override
  Widget build(BuildContext context) {
    return TextField(
      controller: widget.textController,
      decoration: InputDecoration(
        hintText: "hint!",
        prefixIcon: Icon(Icons.search),
        suffixIcon: widget.textController.value.text.isEmpty
            ? null
            : IconButton(
                icon: Icon(Icons.clear, color: Colors.white),
                onPressed: () {
                  setState(() {
                    widget.textController.clear();
                  });
                },
              ),
      ),
      onChanged: (String a) {
        setState(() {});
      },
    );
  }
}
```

## Logs

Run logs:
```
flutter: ══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
flutter: The following assertion was thrown while handling a gesture:
flutter: invalid text selection: TextSelection(baseOffset: 4, extentOffset: 4, affinity:
flutter: TextAffinity.downstream, isDirectional: false)
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #0      TextEditingController.selection= (package:flutter/src/widgets/editable_text.dart:169:7)
flutter: #1      EditableTextState._handleSelectionChanged (package:flutter/src/widgets/editable_text.dart:1216:23)
flutter: #2      RenderEditable._handlePotentialSelectionChange (package:flutter/src/rendering/editable.dart:369:5)
flutter: #3      RenderEditable.selectWordEdge (package:flutter/src/rendering/editable.dart:1466:9)
flutter: #4      _TextFieldState._handleSingleTapUp (package:flutter/src/material/text_field.dart:768:27)
flutter: #5      _TextSelectionGestureDetectorState._handleTapUp (package:flutter/src/widgets/text_selection.dart:944:16)
flutter: #6      TapGestureRecognizer._checkUp.<anonymous closure> (package:flutter/src/gestures/tap.dart:363:49)
flutter: #7      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:182:24)
flutter: #8      TapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:363:11)
flutter: #9      TapGestureRecognizer.acceptGesture (package:flutter/src/gestures/tap.dart:312:7)
flutter: #10     _TransparentTapGestureRecognizer.rejectGesture (package:flutter/src/widgets/text_selection.dart:1146:7)
flutter: #11     GestureArenaManager.sweep (package:flutter/src/gestures/arena.dart:159:26)
flutter: #12     _WidgetsFlutterBinding&BindingBase&GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:222:20)
flutter: #13     _WidgetsFlutterBinding&BindingBase&GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:198:22)
flutter: #14     _WidgetsFlutterBinding&BindingBase&GestureBinding._handlePointerEvent (package:flutter/src/gestures/binding.dart:156:7)
flutter: #15     _WidgetsFlutterBinding&BindingBase&GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:102:7)
flutter: #16     _WidgetsFlutterBinding&BindingBase&GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:86:7)
flutter: #20     _invoke1 (dart:ui/hooks.dart:250:10)
flutter: #21     _dispatchPointerDataPacket (dart:ui/hooks.dart:159:5)
flutter: (elided 3 frames from package dart:async)
flutter:
flutter: Handler: "onTapUp"
flutter: Recognizer:
flutter:   _TransparentTapGestureRecognizer#fa3c6
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════
```

flutter doctor -v
```
[✓] Flutter (Channel dev, v1.8.0, on Mac OS X 10.14.5 18F132, locale en-BR)
    • Flutter version 1.8.0 at /usr/local/Cellar/flutter/dev
    • Framework revision 2fefa8c731 (8 days ago), 2019-07-01 11:33:22 -0700
    • Engine revision 45b66b722e
    • Dart version 2.4.0

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.0)
    • Android SDK at /Users/bernardoferrari/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling
      support)
    • Platform android-29, build-tools 29.0.0
    • Java binary at: /Users/bernardoferrari/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-1/191.5679705/Android
      Studio 3.6 Preview.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 10.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.2.1, Build version 10E1001
    • CocoaPods version 1.7.3

[✓] iOS tools - develop for iOS devices
    • ios-deploy 1.9.4

[!] Android Studio
    • Android Studio at /Users/bernardoferrari/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/191.5675373/Android
      Studio 3.5 Preview.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    • Dart plugin version 191.7830
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)

[✓] Android Studio
    • Android Studio at /Users/bernardoferrari/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-1/191.5679705/Android
      Studio 3.6 Preview.app/Contents
    • Flutter plugin version 36.1.3
    • Dart plugin version 191.7830
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)

[✓] IntelliJ IDEA Ultimate Edition (version 2019.1.3)
    • IntelliJ at /Users/bernardoferrari/Applications/JetBrains Toolbox/IntelliJ
      IDEA Ultimate.app
    • Flutter plugin version 37.0.3
    • Dart plugin version 191.7830

[✓] VS Code (version 1.36.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.2.0

[✓] Connected device (1 available)
    • iPhone Xʀ • 0001D530-E48B-4691-ADD7-E9E95F56B340 • ios •
      com.apple.CoreSimulator.SimRuntime.iOS-12-2 (simulator)

! Doctor found issues in 1 category.
```

### PR DESCRIPTION
**Pull request by perclasson**
#### Reproducible example: [See dartpad](https://dartpad.dartlang.org/0c8069eb160d50440ba2961f8d4c7a05)

For `TextField`s with a suffixIcon, if we tap on the suffixIcon we will get two tap events:
1. The `onTap` gesture will be activated for the `IconButton`, which is clearing the TextField.
2. As we're tapping inside of the `TextField`, we will also have a tap event for the [`TextSelectionGestureDetector`](https://github.com/flutter/flutter/blob/0b8abb4724/packages/flutter/lib/src/widgets/text_selection.dart#L1135) that selects text. 

For the `TextSelectionGestureDetector` to figure out what text to select it uses the [`_textPainter`](https://github.com/flutter/flutter/blob/f15c887c6349b10a0802e7d191cb109e7a70a17d/packages/flutter/lib/src/rendering/editable.dart#L244). Inside of the [`EditableText.selectWordEdge`](https://github.com/flutter/flutter/blob/f15c887c6349b10a0802e7d191cb109e7a70a17d/packages/flutter/lib/src/rendering/editable.dart#L1659) we call `_textPainter.getPositionForOffset`. The `_textPainter` contains the text before we cleared it, thus we are getting invalid positions to select.

To fix this, we're now checking that the new selection is valid before updating it.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/35848

## Tests

I added the following tests:

- Clearing text field with suffixIcon does not cause text selection exception
## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by rami-a — APPROVED** [review]

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -1530,6 +1536,12 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
   }
 
   void _handleSelectionChanged(TextSelection selection, RenderEditable renderObject, SelectionChangedCause cause) {
+    // We return early if the selection is not valid. This can happen when the
+    // text of [EditableText] is updated at the same time as the selection is
```
Comment (guidezpl) [inline packages/flutter/lib/src/widgets/editable_text.dart]: Use /// for highlighting

Comment (perclasson) [reply packages/flutter/lib/src/widgets/editable_text.dart]: This is a [comment](https://dart.dev/guides/language/effective-dart/documentation#do-format-comments-like-sentences) though and not a doc comment. So I'll keep it as it is.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 2]
```diff
@@ -220,6 +220,12 @@ class TextEditingController extends ValueNotifier<TextEditingValue> {
   void clearComposing() {
     value = value.copyWith(composing: TextRange.empty);
   }
+
+  /// Check that the [selection] is valid, and not outside of the bounds of
+  /// [text].
+  bool isSelectionValid(TextSelection selection) {
+    return selection.start <= text.length && selection.end <= text.length;
+  }
```
Comment (pennzht) [inline packages/flutter/lib/src/widgets/editable_text.dart]: Are there any conditions that `selection.start` and `selection.end` must satisfy? Such as `selection.start >= 0` or `selection.start <= selection.end`?

I see you are extracting an existing function here; I'm just wondering if this is the correct logic.

**Review by pennzht — APPROVED** [review]

**Review by HansMuller — APPROVED** [review]
LGTM however @justinmc should take a look as well

**Review by justinmc — APPROVED** [review]
LGTM 👍 Thanks for fixing this, it's been around for awhile!

**Pull request merged**